### PR TITLE
#9024 Number formatting should be internationalized

### LINKED ISSUE #9024: Number formatting should be internationalized
A user tweet claimed that a lack of number localization made MetaMask very difficult for them to use: We use commas for thousands, their region used periods.

This should be derived from the user's region code.

### PR DESCRIPTION
**Pull request by ebaker**
### Scope

- [x] Updated `formatCurrency` to accept `currentLocale` and use `Intl.NumberFormat` for localized currency formatting
- [x] Added `@formatjs/intl-numberformat` polyfill, copying locale-data for locales listed in `app/_locales/index.json`
- [x] Added `format-currency-data.json` which is generated from native browser `Intl.NumberFormat` for locales listed in `app/_locales/index.json` format of USD & EUR
- [x] Updated `formatCurrency` calls & unit tests using `formatCurrency` to be localized

### Notes

- I did not remove `currency-formatter` package dependency as it was also being used for a list of fiat currencies. Happy to remove this and replace it with [openexchangerates.org - currencies.json](https://openexchangerates.org/api/currencies.json) or [currency-codes](https://www.npmjs.com/package/currency-codes) if one of these is preferred.
- The formatted currency data generated from Firefox & Chrome for locales `bg` & `es-419` did not match the format from `@formatjs/int-numberformat` polyfill. In researching, the polyfill format seemed accurate and I updated the test data.
- I used formatting for locale `nb` in place of `no` as `@formatjs/int-numberformat` polyfill did not contain locale-data for `no`.

Happy to make any changes or updates. Thanks!

Fixes #9024

### CODE REVIEW

**Comment by brad-decker** [comment]
Greetings @ebaker 👋 ,

First off thank you for this contribution! At first glance, the code looks good -- I'll need to dig in a little bit to provide feedback to help get this merged. I'll be doing that today and Monday, and will be circling up with other team members to think through the issue with Mozilla lint. The build is failing because importing the additional locale data required for internationalization pushes the `ui.js` bundle (previously around 3.65mb) to > 4mb which is a limit imposed by Mozilla lint. 

I think @Gudahtt, who was giving you feedback and pointers on the issue, did something fancy with importing locale data for our previous usage of `@formatjs` -- perhaps we can tack onto this. I'll talk with the team and dig into the code to see if we can get this resolved.

Thanks Again!

**Comment by ebaker** [comment]
Hey  @brad-decker ,

Thanks for the positivity in your quick look of the PR 😄  After seeing the `mozilla-lint` job failing, I did some investigation and realized the way @Gudahtt  imported the data for `Intl.RelativeTimeFormat`. Very clever @Gudahtt ! 👏 

I attempted to mirror @Gudahtt 's pattern for `Intl.NumberFormat`, however the latest and more recent versions of `@formatjs/intl-numberformat` no longer expose the locale data as `.json` file - instead `@formatjs/intl-numberformat` exposes `.js` files to be dynamically imported. This is also how the more recent versions of `@formatjs/int-relativetimeformat` are being exported as well FYI.

I tried using `@formatjs/intl-numberformat@4.2.10`, which is the last version that exposes the `json` files for the locale data, however in my testing the output of this version of the library some locales used in metamask were not consistent with the output of `NumberFormat` in current versions of my browser.

I pushed a naive fix to this branch that extracts the locale data json from the `.js` file exported `@formatjs/intl-numberformat` via regex and fixed the version of `@formatjs/intl-numberformat` to the current version. The code that generates the js files can be found here: https://github.com/formatjs/formatjs/blob/7fe75feb99a80c49102968d0326a43c101eab5e1/packages/intl-numberformat/scripts/cldr.ts#L51

I realize the fix I implemented is not ideal, a more sensible alternative may be to allow `eval()` of these scripts.

I'll leave the polyfill imports in their current state so that the team can discuss the approach they'd like to take and tack onto this PR.

Happy to make any changes/suggestions to code outside of the loading of the polyfill. Cheers!

**Comment by darkwing** [comment]
Hello @ebaker!  Would you have some time to rebase?  We recently added prettier so conflicts are likely just formatting changes.  Thank you!

**Comment by Gudahtt** [comment]
The use in this PR of the `Intl.NumberFormat` API might work in all of our supported browsers without a polyfill :thinking: I'm going to test this.

In the issue I had suggested that we'd want to use some option that wasn't supported by all browsers we support (many options weren't introduced until Chrome v77), but looking back on that now, I'm not sure we need any of those.

**Comment by Gudahtt** [comment]
I have rebased this branch onto develop by stepping through each commit one-by-one, resolving conflicts and running `prettier` on each one. Most of the conflicts were lint-related, but there were a few major changes relating to swaps (either new code introduced or substantially refactored). I just applied the same approach used elsewhere to update these parts (i.e. ensure the locale is always passed into `formatCurrency`). It all seems to work!

I tested this on our minimum Firefox and Chromium versions without the polyfill. The option `currencyDisplay: 'narrowSymbol'` is not supported on Chromium v65 unfortunately (probably not until v77 I'm guessing, though I had trouble finding confirmation of this).

The default behaviour of `currencyDisplay` will show "US$" instead of "$" in some locales, which looks pretty redundant next to the currency symbol.  For example:
doubleusd

So... given that, perhaps we should keep the polyfill until we increase the minimum Chrome version to v77.

All of the non-polyfill-related code looks good to me, but I haven't reviewed the polyfill changes in detail yet.
 I'll work on that next.

**Comment by Gudahtt** [comment]
@ebaker would you be willing to sign our CLA? We won't be able to accept this contribution unless you do, and I'd _really_ like to see this merged. The instructions on this are [here](https://github.com/MetaMask/metamask-extension/pull/9311#issuecomment-721318958).

Sorry for the long delay in reviewing this! I'll ensure we're more prompt in replying from now on.

**Comment by ebaker** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Gudahtt** [comment]
I'm having trouble getting this to work on Node.js v10 :confused: Specifically the unit tests. We may have to block this on updating to Node.js v12 (which is happening pretty soon anyway).

**Comment by Gudahtt** [comment]
I've rebased this on `develop`, and everything seems to be working now! :tada: 

In testing everything seems to be working, though I did notice that a lot of numbers still aren't formatted properly yet (e.g. the main ETH balance. I'm not sure whether that should be a blocker or not :thinking:. I'm concerned that if we have numbers formatted differently throughout the application, it could lead to more confusion (e.g. if a decimal point is mistaken for a numeric separator or vice versa). I'll look into the effort involved here.

File: ui/app/helpers/utils/i18n-helper.js [hunk 1]
```diff
@@ -108,9 +108,44 @@ export async function fetchLocale(localeCode) {
...
+  if (
+    Intl.NumberFormat &&
+    typeof Intl.NumberFormat.__addLocaleData === 'function' &&
+    !numberFormatLocaleData.has(languageTag)
+  ) {
+    const localeDataScript = await fetchNumberFormatLocaleData(languageTag);
+    if (!localeDataScript) {
+      return;
+    }
+    const localeDataRegex = `\\/\\* @generated \\*\\/
```
Comment (Gudahtt) [inline ui/app/helpers/utils/i18n-helper.js:124]: Wow O_O. I think I see why you did this, but this is pretty unpleasant. I'm going to see if I can get a dynamic require working here instead.

Comment (Gudahtt) [reply ui/app/helpers/utils/i18n-helper.js]: Done in bd58656ea96d8c87ca59448331eec5af77e1b292 - let me know what you think!

**Comment by Gudahtt** [comment]
We've decided to audit the extension to find all displayed numbers, and ensure they're all being formatted consistently. The concern was that inconsistent formatting might introduce more confusion, and result in users misunderstanding where decimal points are and so on.

**Pull request closed without merging**